### PR TITLE
Improved the ThreadPool

### PR DESCRIPTION
**Pull request by emilwestergren**

### CODE REVIEW

File: source/base/ThreadPool.ooc [hunk 1]
```diff
@@ -1,33 +1,65 @@
...
+	}
+	run: override func {
+		this _execute()
+		this _mutex lock()
+		this _finished = true
+		if (this _freeOnCompletion) {
+			this _mutex unlock()
+			this free()
+		}
+		else {
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: } else {

**Comment by emilwestergren** [comment]
Updated.

File: source/base/ThreadPool.ooc [hunk 2]
```diff
@@ -1,33 +1,63 @@
...
+	_freeOnCompletion := false
+	init: func (body: Func) { super(body) }
 	free: override func {
-		this _finishedCondition free()
-		super()
+		this _mutex lock()
+		if (!this _finished) {
+			this _freeOnCompletion = true
+			this _mutex unlock()
+		}	else {
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: There's a tab after the `}`.

File: source/base/ThreadPool.ooc [hunk 3]
```diff
@@ -1,33 +1,63 @@
...
+		}
+	}
+	run: override func {
+		this _execute()
+		this _mutex lock()
+		this _finished = true
+		if (this _freeOnCompletion) {
+			this _mutex unlock()
+			this free()
+		}	else {
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: And here.

File: source/base/ThreadPool.ooc [hunk 4]
```diff
@@ -47,45 +72,33 @@ ThreadPool: class {
...
-				this _activeJobs -= 1
-				if(this _activeJobs == 0)
-					this _allFinishedCondition broadcast()
-				this _mutex unlock()
-			} else {
-				this _newJobCondition wait(this _mutex)
-				this _mutex unlock()
-			}
+			this _activeJobs -= 1
+			if(this _activeJobs == 0)
```
Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: space after `if`

**Comment by emilwestergren** [comment]
Updated.

**Comment by emilwestergren** [comment]
Rebased.

**Pull request merged**